PrebidMobile/PrebidMobileCore/src/main/java/org/prebid/mobile/core/Prebid.java

```diff
@@ -182,11 +193,52 @@ private static Object callMethodOnObject(Object object, String methodName, Objec
 
     private static void handleMoPubKeywordsUpdate(Object adViewObj, String adUnitCode, Context context) {
         ArrayList<Pair<String, String>> keywordPairs = BidManager.getKeywordsForAdUnit(adUnitCode, context);
-
         if (keywordPairs != null && !keywordPairs.isEmpty()) {
             StringBuilder keywords = new StringBuilder();
+            // check if the bid should be load in demand sdk or MoPub ad view/interstitial object
+            boolean bidShouldBeLoadedInDemandSdk = false;
+            String bidder = "";
             for (Pair<String, String> p : keywordPairs) {
-                keywords.append(p.first).append(":").append(p.second).append(",");
+                if (PREBID_CREATIVE_LOAD_TYPE.equals(p.first)) {
+                    if (PREBID_DEMAND_SDK.equals(p.second)) {
+                        bidShouldBeLoadedInDemandSdk = true;
+                    }
+                }
+                if (PREBID_BIDDER.equals(p.first)) {
+                    bidder = p.second;
+                }
+            }
+            if (bidShouldBeLoadedInDemandSdk) {
+                // pass bids that require demand sdk only when rendering is enabled
+                // save the cache id locally on the ad object for later use
+                Class mediation_adapter = getClassFromString(PREBID_MOPUB_CUSTOM_EVENT_BANNER);
+                if (mediation_adapter != null && PrebidDemandSettings.isDemandEnabled(bidder)) {
+                    for (Pair<String, String> p : keywordPairs) {
+                        keywords.append(p.first).append(":").append(p.second).append(",");
+                        if (PREBID_CACHE_ID.equals(p.first) || PREBID_BIDDER.equals(p.first)) {
+                            Map<String, Object> localExtras = (Map) callMethodOnObject(adViewObj, "getLocalExtras");
```

Comment:
Is there a reason the localExtras need to be set in the loop? Maybe if you pull in cache id above then we don't even need the loop here.

Reply:
MoPub returns a copy of the localExtras not the original object, so the copy of data is needed here.
What do you mean by pull in cache id above?